google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/internal/AssigningSubscriber.java

```diff
@@ -0,0 +1,123 @@
+import com.google.cloud.pubsublite.internal.wire.Assigner;
+import com.google.cloud.pubsublite.internal.wire.AssignerFactory;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.grpc.StatusException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class AssigningSubscriber extends ProxyService implements Subscriber {
+  private final PartitionSubscriberFactory subscriberFactory;
+
+  private final CloseableMonitor monitor = new CloseableMonitor();
+
+  @GuardedBy("monitor.monitor")
+  private final Map<Partition, Subscriber> liveSubscriberMap = new HashMap<>();
+
+  @GuardedBy("monitor.monitor")
+  private final List<Subscriber> stoppingSubscribers = new ArrayList<>();
+
+  @GuardedBy("monitor.monitor")
+  private boolean shutdown = false;
+
+  public AssigningSubscriber(
+      PartitionSubscriberFactory subscriberFactory, AssignerFactory assignerFactory)
+      throws StatusException {
+    this.subscriberFactory = subscriberFactory;
+    Assigner assigner = assignerFactory.New(this::handleAssignment);
+    addServices(assigner);
+  }
+
+  @Override
+  protected void start() {}
+
+  @Override
+  protected void stop() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      shutdown = true;
+      liveSubscriberMap.values().forEach(subscriber -> subscriber.stopAsync().awaitTerminated());
+      liveSubscriberMap.clear();
+      stoppingSubscribers.forEach(Subscriber::awaitTerminated);
+    }
+  }
+
+  @Override
+  protected void handlePermanentError(StatusException error) {
+    stop();
+  }
+
+  private void handleAssignment(Set<Partition> assignment) {
+    try {
+      try (CloseableMonitor.Hold h = monitor.enter()) {
+        if (shutdown) return;
+        Set<Partition> livePartitions = ImmutableSet.copyOf(liveSubscriberMap.keySet());
+        for (Partition partition : livePartitions) {
+          if (!assignment.contains(partition)) {
+            stopSubscriber(liveSubscriberMap.remove(partition));
+          }
+        }
+        for (Partition partition : assignment) {
+          if (!liveSubscriberMap.containsKey(partition)) startSubscriber(partition);
+        }
+      }
+    } catch (StatusException e) {
+      onPermanentError(e);
+    }
+  }
+
+  @GuardedBy("monitor.monitor")
+  private void startSubscriber(Partition partition) throws StatusException {
+    checkState(!liveSubscriberMap.containsKey(partition));
+    Subscriber subscriber = subscriberFactory.New(partition);
+    subscriber.addListener(
+        new Listener() {
+          @Override
+          public void failed(State from, Throwable failure) {
+            onPermanentError(toCanonical(failure));
+          }
+
+          @Override
+          public void terminated(State from) {
+            stoppingSubscribers.remove(subscriber);
```

Comment:
I'm a bit worried about having multiple "stopping subscribers" for the same partition, on assignment churn. WDYT?

Reply:
I'm not particularly, although this code handles that case since its checking object identity not partition number. The subsequent subscribers will force-evict older ones, so will just speed up them shutting down by failing their streams with an error, so it should be okay.